### ISSUE TITLE
Incremental learning

### ISSUE THREAD

**Issue by 0wwafa** [ggerganov/llama.cpp#7511]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x ] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [x ] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [x ] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Feature Description

I wish there would gbe an option in main and server, to let the model learn (fine tune) while chatting interactvely.
Every XX minutes or every XX tokens it would save the checkpoint...

Later on, the model and checkpoint could be merged and tested.

# Motivation
To be able to teach a model iteractively like teaching to a child.

# Possible Uses

Countless... from experimenting to see in real time how a model learns.

**Comment by mirek190** [comment]
Such idea is great but merge knowledge from working memory (context limit) is very compute demand in current state .... maybe that's why we have to sleep 8 hours to merge working memory - is also very compute even for human. :)

**Comment by 0wwafa** [comment]
My personal aim in the ai world is to try to create (or use) and define the best ais based on size/hardware demands, speed.
I don't say the paradigm must shift, research is research, but I think that optimizing things would make text generating ai enjoyable even on an old laptop or edge small device. I made some tests, and as of now a few ais shine even when heavily quantized:
**Mistral-7b-Instruct-v0.3** Very good.
WizardLM-2-7B   Not so bad.
**Tinybra_13B**  The best so far.
and without any finetuning.
They tun between 6 t/s and 3t/s on a corei7 without a usable GPU.

As per the "host" program, llama.cpp seems the more optimized and slim of them all.

The great thing to do would be to create an "ignorant genius": an A.I. with a great logic and problem solving and mathematical skills, so with the "tools" to learn anything else and a small footprint (think of it as a child).

I noticed that today, A.I. is a powerplay and "brute force" reigns... tons of data in datasets which are often of medium to low quality.
I wish to do the opposite.
Also I have no real computing resources to do those things so I will need help, but I have a clear idea on how to make things work.
Since I am still learning, I'm sorry if I asked something that sounded silly.
